packages/admin/admin/src/router/PromptHandler.tsx

```diff
@@ -5,13 +5,94 @@ import { Prompt } from "react-router";
 import { RouterContext } from "./Context";
 
-interface IMessages {
-    [id: string]: (location: History.Location, action: History.Action) => boolean | string;
+interface PromptHandlerState {
+    showConfirmationDialog: boolean;
+    message: string;
+    callback?: (ok: boolean) => void;
+}
+export interface PromptHandlerApi {
+    showDialog: (message: string, callback: (ok: boolean) => void) => void;
+}
+function InnerPromptHandler({
+    registeredMessages,
+    saveActions,
+    apiRef,
+}: {
+    registeredMessages: React.MutableRefObject<PromptMessages>;
+    saveActions: React.MutableRefObject<SaveActions>;
+    apiRef: React.MutableRefObject<PromptHandlerApi | undefined>;
+}) {
+    const [state, setState] = React.useState<PromptHandlerState>({
+        showConfirmationDialog: false,
+        message: "",
+        callback: undefined,
+    });
+    if (apiRef)
+        apiRef.current = {
+            showDialog: (message: string, callback: (ok: boolean) => void) => {
+                setState({
+                    showConfirmationDialog: true,
+                    message,
+                    callback,
+                });
+            },
+        };
+
+    const promptMessage = (location: History.Location, action: History.Action): boolean | string => {
+        let ret: boolean | string = true;
+        Object.keys(registeredMessages.current).forEach((id) => {
+            if (location.pathname.startsWith(registeredMessages.current[id].path)) {
+                // allow transition if location is below path where prompt was rendered
+                return true;
+            }
+            const message = registeredMessages.current[id].message(location, action);
+            if (message !== true) {
+                ret = message;
+                return false;
+            }
+        });
+        return ret;
+    };
+
+    const handleClose = async (action: PromptAction) => {
+        let allowTransition: boolean;
+        if (Object.keys(saveActions.current).length > 0 && action === PromptAction.Save) {
+            const results: Array<SaveActionSuccess> = await Promise.all(Object.keys(saveActions.current).map((id) => saveActions.current[id]()));
+            allowTransition = results.every((saveActionSuccess) => saveActionSuccess);
+        } else {
+            allowTransition = action === PromptAction.Discard;
+        }
+        if (state.callback) {
+            state.callback(allowTransition);
+        }
+        setState({
+            showConfirmationDialog: false,
+            message: "",
+            callback: undefined,
+        });
+    };
+
+    return (
+        <>
+            <RouterConfirmationDialog
+                isOpen={state.showConfirmationDialog}
+                message={state.message}
+                handleClose={handleClose}
+                showSaveButton={Object.keys(saveActions.current).length > 0}
+            />
+            <Prompt when={true} message={promptMessage} />
```

Comment:
Why don't you do the check here and use the `when` prop instead of returning true in `promptMessage`? Is this not possible?

Reply:
what check?

Reply:
The check to show the prompt

<img width="395" alt="image" src="https://user-images.githubusercontent.com/48853629/228747860-7cbecf7f-49ab-4fbe-b917-34e7b503573b.png">


https://v5.reactrouter.com/core/api/Prompt

Reply:
because message is called on demand (when location changes) and this allows us to check registeredMessages (which are non reactive!) on demand